File: internal/adapters/http/dto/sync.go
package dto

import (
	"github.com/longregen/alicia/internal/domain/models"
)

type SyncMessageRequest struct {
	LocalID        string `json:"local_id" msgpack:"localId"`
	SequenceNumber int    `json:"sequence_number" msgpack:"sequenceNumber"`
	PreviousID     string `json:"previous_id,omitempty" msgpack:"previousId,omitempty"`
	Role           string `json:"role" msgpack:"role"`
	Contents       string `json:"contents" msgpack:"contents"`
	CreatedAt      string `json:"created_at" msgpack:"createdAt"`
	UpdatedAt      string `json:"updated_at,omitempty" msgpack:"updatedAt,omitempty"`
}

type SyncRequest struct {
	Messages []SyncMessageRequest `json:"messages" msgpack:"messages"`
}

type SyncedMessage struct {
	LocalID  string           `json:"local_id" msgpack:"localId"`
	ServerID string           `json:"server_id" msgpack:"serverId"`
	Status   string           `json:"status" msgpack:"status"`
	Message  *MessageResponse `json:"message,omitempty" msgpack:"message,omitempty"`
	Conflict *ConflictDetails `json:"conflict,omitempty" msgpack:"conflict,omitempty"`
}

type ConflictDetails struct {
	Reason        string           `json:"reason" msgpack:"reason"`
	ServerMessage *MessageResponse `json:"server_message,omitempty" msgpack:"serverMessage,omitempty"`
	Resolution    string           `json:"resolution" msgpack:"resolution"`
}

type SyncResponse struct {
	SyncedMessages []SyncedMessage `json:"synced_messages" msgpack:"syncedMessages"`
	SyncedAt       string          `json:"synced_at" msgpack:"syncedAt"`
}

type SyncStatusResponse struct {
	ConversationID string `json:"conversation_id" msgpack:"conversationId"`
	PendingCount   int    `json:"pending_count" msgpack:"pendingCount"`
	SyncedCount    int    `json:"synced_count" msgpack:"syncedCount"`
	ConflictCount  int    `json:"conflict_count" msgpack:"conflictCount"`
	LastSyncedAt   string `json:"last_synced_at,omitempty" msgpack:"lastSyncedAt,omitempty"`
}

func ToSyncedMessage(msg *models.Message) SyncedMessage {
	status := string(msg.SyncStatus)
	if status == "" {
		status = "synced"
	}

	return SyncedMessage{
		LocalID:  msg.LocalID,
		ServerID: msg.ServerID,
		Status:   status,
		Message:  (&MessageResponse{}).FromModel(msg),
	}
}

func ToSyncedMessageWithConflict(localID string, reason string, serverMsg *models.Message) SyncedMessage {
	var serverMsgResp *MessageResponse
	if serverMsg != nil {
		serverMsgResp = (&MessageResponse{}).FromModel(serverMsg)
	}

	return SyncedMessage{
		LocalID: localID,
		Status:  "conflict",
		Conflict: &ConflictDetails{
			Reason:        reason,
			ServerMessage: serverMsgResp,
			Resolution:    "manual",
		},
	}
}

File: frontend/src/types/sync.ts
import { Message } from './models';

// Wire format types using camelCase to match Go msgpack serialization
// The Go backend uses msgpack tags like `msgpack:"syncedMessages"`

export interface SyncMessageRequest {
  localId: string;
  sequenceNumber: number;
  previousId?: string;
  role: string;
  contents: string;
  createdAt: string;
  updatedAt?: string;
}

export interface SyncRequest {
  messages: SyncMessageRequest[];
}

export interface ConflictDetails {
  reason: string;
  serverMessage?: MessageResponse;
  resolution: string;
  /** When the local version was last modified */
  localModifiedAt?: string;
  /** When the server version was last modified */
  serverModifiedAt?: string;
  /** The type of conflict (content, metadata, etc.) */
  conflictType?: 'content' | 'metadata' | 'both';
  /** Additional context about the conflict */
  description?: string;
}

export interface SyncedMessage {
  localId: string;
  serverId: string;
  status: 'synced' | 'conflict';
  message?: MessageResponse;
  conflict?: ConflictDetails;
}

export interface SyncResponse {
  syncedMessages: SyncedMessage[];
  syncedAt: string;
}

export interface SyncStatusResponse {
  conversationId: string;
  pendingCount: number;
  syncedCount: number;
  conflictCount: number;
  lastSyncedAt?: string;
}

// MessageResponse is the wire format for messages (camelCase)
// This matches the Go msgpack serialization
export interface MessageResponse {
  id: string;
  conversationId: string;
  sequenceNumber: number;
  previousId?: string;
  role: 'user' | 'assistant' | 'system';
  contents: string;
  createdAt: string;
  updatedAt: string;
  localId?: string;
  serverId?: string;
  syncStatus?: 'pending' | 'synced' | 'conflict';
}

// Helper to convert wire format (camelCase) to domain model (snake_case)
export function messageResponseToMessage(response: MessageResponse): Message {
  return {
    id: response.id,
    conversation_id: response.conversationId,
    sequence_number: response.sequenceNumber,
    previous_id: response.previousId,
    role: response.role,
    contents: response.contents,
    created_at: response.createdAt,
    updated_at: response.updatedAt,
    local_id: response.localId,
    server_id: response.serverId,
    sync_status: response.syncStatus,
  };
}

export interface SyncState {
  isSyncing: boolean;
  lastSyncTime: Date | null;
  syncError: string | null;
}
